File: scala/grade-school/src/main/scala/School.scala
import scala.collection.immutable.SortedMap

class School {
  private var _db: Map[Int, Seq[String]] =
    SortedMap.empty[Int, Seq[String]]
      .withDefaultValue(Nil)

  def db = _db

  def add(name:String, grade:Int) = {
    _db += grade -> (_db(grade) :+ name)
  }

  def grade = _db

  def sorted = _db.mapValues(_.sorted)
}

File: haskell/grade-school/src/School.hs
module School (School, add, empty, grade, sorted) where

import Data.List (sort)
import qualified Data.Map as M

type School = M.Map Int [String]
type Grades = [(Int, [String])]

add :: Int -> String -> School -> School
add g name = M.insertWith (flip (++)) g [name]

empty :: School
empty = M.empty

grade :: Int -> School -> [String]
grade = M.findWithDefault []

sorted :: School -> Grades
sorted = M.toList . M.map sort
